### ISSUE TITLE
lint assert!(true / false)

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#3575]
`assert!(true);` will be optimized out by the compiler
`assert!(false);` should probably be replaced by a `panic!()`.

**Comment by Arkweid** [comment]
Take this one :)

**Comment by Arkweid** [comment]
PR added :)
@matthiaskrgr can you check? Especially the error messages, I'm not sure what the description is as it should be

**Comment by flip1995** [comment]
After #3582 was merged this lint can be enhanced by suggestions for `assert(true/false)`:
- suggest `""`/`"panic!()"` (should be easy, but some weird things were going on: https://github.com/rust-lang/rust-clippy/pull/3582#discussion_r246656947)
- if it is a `assert!(false, "_")` call, suggest `panic!("_")` instead of just `panic!()` (advanced)

**Comment by Lythenas** [comment]
I would like to try and implement the first suggestion (`assert(true)` -> `""` and `assert(false)` -> `panic!()`). And after maybe the second suggestion too.

**Comment by flip1995** [comment]
Just for clarification:

> `assert(true)` -> `""` and `assert(false)` -> `panic!()`

suggesting `""` instead of assert doesn't make sense. Then I saw, that I recommended to do this:

> suggest `""`/`"panic!()"`

What I meant with that is, to suggest removing the assert completely if it is `assert!(true)` and to suggest `panic!()` if it is `assert!(false)`.

**Comment by Lythenas** [comment]
Right. I didn't think that far. Also I notices the first part is already implemented so I will start on the second part `assert!(false, "msg")` -> `panic/unreachable!("msg")`

**Comment by Lythenas** [comment]
I will need some help here.

I found
https://github.com/rust-lang/rust-clippy/blob/54bf4ffd626970e831bb80c037f804a3b3450835/clippy_lints/src/write.rs#L324-L441

But it uses `EarlyContext` and the `assertions_on_constants` is a `LateLintPass`. Is there still some way to reuse this? Or am I look at this the wrong way?

**Comment by flip1995** [comment]
> assertions_on_constants is a LateLintPass

Yes, that's why I wrote (advanced) behind this. What you need to do is to look at the expanded code. You can do this with the rustc flag `-Zunpretty=hir`.

For example:
```rust
fn main() {
    assert!(false, "some {} things", "String");
}
```
expands to
```rust
#[prelude_import]
use ::std::prelude::v1::*;
#[macro_use]
extern crate std;
fn main() {
              match { let _t = !false; _t } {
                  true => {
                      {
                          ::std::rt::begin_panic_fmt(&<::core::fmt::Arguments>::new_v1(&["some ",
                                                                                         " things"],
                                                                                       &match (&"String",)
                                                                                            {
                                                                                            (arg0,)
                                                                                            =>
                                                                                            [<::core::fmt::ArgumentV1>::new(arg0,
                                                                                                                            ::core::fmt::Display::fmt)],
                                                                                        }),
                                                     &("main.rs", 2u32, 5u32))
                      }
                  }
                  _ => { }
              };
          }
```

You then have to match the expanded construct. Something similar was done before in `clippy_lints/src/format.rs`, so you may be able to reuse this code. It would also be great to have a `utils` function that preduces a suggestion for this `core::fmt::Arguments::new_v1` things, since this is not trivial. But a first step would be to only produce suggestions for `assert!`s without format args. (`assert!(false, "no fmt args")`), since something similar is already done in  `clippy_lints/src/format.rs`.

**Comment by CAD97** [comment]
[The implementation](https://github.com/rust-lang/rust-clippy/blob/a71211d0b52c01f1b37fe544e0e13fd1bdc31979/clippy_lints/src/assertions_on_constants.rs#L32-L84) now evaluates the constant and emits different help messages based on whether the assertion is always `true` or `false`. I think that's sufficient to close this issue.

With #11966 fixed (not linting when the `assert!` is always constant evaluated), `assert!(TRUE)` should be replaced with `const _: () = assert!(TRUE);` (or similar) to ensure constant evaluation, and `assert!(FALSE)` should (almost[^almost]) always be a `panic!`ing macro instead.

[^almost]: The exception being "fickle" consts that vary based on cfg and/or lib versions, which could validly only want a runtime panic.
